prometheus/graphite/bridge.go

```diff
@@ -0,0 +1,230 @@
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+
+	dto "github.com/prometheus/client_model/go"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+)
+
+// Config defines the graphite bridge config.
+type Config struct {
+	// The url to push data to. Required.
+	URL string
+
+	// The interval to use for pushing data to Graphite. Defaults to 15 seconds.
+	Interval time.Duration
+
+	// The Gatherer to use for metrics. Defaults to prometheus.DefaultGatherer.
+	Gatherer prometheus.Gatherer
+
+	// The logger that messages are written to. Defaults to log.Base().
+	Logger log.Logger
+
+	// The prefix for your graphite metric. Defaults to empty string.
+	Prefix string
+}
+
+// Bridge pushes metrics to the configured graphite server.
+type Bridge struct {
+	url      string
+	interval time.Duration
+	prefix   string
+
+	g      prometheus.Gatherer
+	logger log.Logger
+}
+
+// NewBridge returns a pointer to a new Bridge struct.
+func NewBridge(c *Config) (*Bridge, error) {
+	b := &Bridge{}
+
+	if c.URL == "" {
+		return nil, errors.New("graphite bridge: no url given")
+	}
+	b.url = c.URL
+
+	if c.Gatherer == nil {
+		b.g = prometheus.DefaultGatherer
+	} else {
+		b.g = c.Gatherer
+	}
+
+	if c.Logger == nil {
+		b.logger = log.Base()
+	} else {
+		b.logger = c.Logger
+	}
+
+	if c.Prefix != "" {
+		b.prefix = c.Prefix
+	}
+
+	var z time.Duration
+	if c.Interval == z {
+		b.interval = 15 * time.Second
+	} else {
+		b.interval = c.Interval
+	}
+
+	return b, nil
+}
+
+// Run starts the event loop that pushes Prometheus metrics to Graphite at the
+// configured interval.
+func (b *Bridge) Run(ctx context.Context) {
+	ticker := time.NewTicker(b.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if err := b.Push(); err != nil {
+				b.logger.Errorf("%v", err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Push pushes Prometheus metrics to the configured Graphite server.
+func (b *Bridge) Push() error {
+	mfs, err := b.g.Gather()
+	if err != nil {
+		return err
+	}
+
+	now := time.Now().Unix()
+	// TODO: Write directly to conn?
+	buf, err := toReader(mfs, b.prefix, now)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Should we expose a deadline to the user?
+	conn, err := net.Dial("tcp", b.url)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	_, err = io.Copy(conn, buf)
+	return err
+}
+
+// TODO: Do we want to allow partial writes? i.e., Skip a failed metric, but
+// try to write all metrics?
+func toReader(mfs []*dto.MetricFamily, prefix string, now int64) (*bytes.Buffer, error) {
+	vec := Vector{
+		Vector: expfmt.ExtractSamples(&expfmt.DecodeOptions{
+			Timestamp: model.Time(now),
+		}, mfs...),
+		prefix: prefix,
+	}
+
+	return bytes.NewBufferString(vec.String()), nil
+}
+
+// Vector wraps the type model.Vector to implement the String() method for
+// Graphite formatting.
+type Vector struct {
```

Comment:
Not sure if this is really the best approach. It's a lot of "design fluff" to create all those separate types to wrap the orginial model and then funnel everything through re-implemented String() methods. It's only a one-time use, after all. It's also not very resource-efficient to put everything into strings, join them into other strings, to finally write them into a Writer anyway.

How about creating a set of normal functions that convert things and write the result into the Writer directly?